src/viper/pretty.ml

```diff
@@ -206,13 +207,19 @@ and pp_stmt' ppf = function
       fprintf ppf "@[%a@]" (pp_print_list ~pp_sep:comma pp_exp) args
     in
     fprintf ppf ")"
-  | LabelS (_, _) -> failwith "LabelS?"
+  | LabelS lbl ->
+      fprintf ppf "@[label %s@]" lbl.it
+  | GotoS lbl ->
```

Comment:
I see that you are generating goto statements which look redundant. What' your main motivation for adding them to the translation?

Reply:
`goto` is used to translate return statements, consider e.g.

```
  public func abs(n : Int) : async Int {
    if (n < 0) {
      return (0 - n);
    };
    return n;
  };
```

Here we have early exit in the `if` branch, which I decided to encode with a `goto` to a label at the end of the function body. Generated Viper code:

```
method abs($Self: Ref, n: Int)
     returns ($Res: Int)
    requires $Perm($Self)
    requires $Inv($Self)
    ensures $Perm($Self)
    ensures $Inv($Self)
    { 
      if ((n < 0))
         { 
           $Res := (0 - n)
           goto $Ret 
         }
      $Res := n
      goto $Ret
      label $Ret 
    }
```

Reply:
Hmm, if Viper doesn't support early exit via return then this seems sensible. The other alternative is to disallow early return from the Motoko fragment (returns are optional and a method can just exit with a final value).

Reply:
@aterga is this document https://github.com/viperproject/silver/blob/master/documentation/syntax/sil-syntax.txt current? It hasn't been touched in a decade and I (perhaps incorrectly) recall having to read the scala parser to figure out what the grammar of .vpr files actually is...

Reply:
I don't this that document is current. 

Viper does have support for arbitrary goto expressions, so even loops could theoretically be encoded via goto, I think. I recall that it's even possible to annotate Viper gotos with invariants (exactly for supporting loops).

Using gotos to model early return statements should work, but I don't have experience with that. I think no goto invariants would be needed in this case, as no new looping would be introduced to the control flow.

As Claudio says, I'd also recommend not spending the time implementing generic early return statements as that may be involved.

Feel free to rewrite the Motoko examples you wirk with so they do not need to return early.

Reply:
> I'd also recommend not spending the time implementing generic early return statements as that may be involved.

They seem to work as implemented. What am I missing?